File: src/challenges/go/problems/add_binary_test.go
// https://leetcode.com/problems/add-binary

package problems

import (
	"strings"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type AddBinary struct{}

func TestAddBinary(t *testing.T) {
	gen := core.TestSuite[AddBinary]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("11").Param("1").Result("100")
	}).Add(func(tc *core.TestCase) {
		tc.Param("1010").Param("1011").Result("10101")
	}).Add(func(tc *core.TestCase) {
		tc.Param("1101010").Param("111").Result("1110001")
	}).Run(t)
}

func (AddBinary) Solution(a string, b string) string {
	if len(a) > len(b) {
		b = strings.Repeat("0", len(a)-len(b)) + b
	} else {
		a = strings.Repeat("0", len(b)-len(a)) + a
	}

	result := make([]byte, len(a))
	carry := byte(0)
	for i := len(a) - 1; i >= 0; i-- {
		temp := a[i] - '0' + b[i] - '0' + carry
		bit := byte(temp % 2)
		carry = byte(temp / 2)
		result[i] = bit + '0'
	}

	if carry == 1 {
		result = append([]byte{'1'}, result...)
	}

	return string(result)
}

File: src/challenges/python/problems/add_binary_test.py
# https://leetcode.com/problems/add-binary

from core.problem_base import *

class AddBinary(ProblemBase):
    def Solution(self, a: str, b: str) -> str:
        if len(a) > len(b):
            b = b.rjust(len(a), '0')
        else:
            a = a.rjust(len(b), '0')

        result = []
        carry = 0
        for i in range(len(a) - 1, -1, -1):
            temp = int(a[i]) + int(b[i]) + carry
            bit = str(temp % 2)
            carry = temp // 2
            result.insert(0, bit)

        if carry == 1:
            result.insert(0, '1')

        return ''.join(result)

if __name__ == '__main__':
    TestGen(AddBinary) \
        .Add(lambda tc: tc.Param("11").Param("1").Result("100")) \
        .Add(lambda tc: tc.Param("1010").Param("1011").Result("10101")) \
        .Add(lambda tc: tc.Param("1101010").Param("111").Result("1110001")) \
        .Run()
